### DISCUSSION TITLE
Using Javascript class objects

### DISCUSSION THREAD

**Discussion by dainbrump** [SilentVoid13/Templater#767] [Help] [answered]
First... Templater is beyond amazing as a plugin for Obsidian. My only concern is that if I want to build some fairly extensive templates with a bunch of custom functions, it appears that I have to create one js file for each function and export that. This can be really messy after a while.

I have searched and experimented but I cannot seem to find an easy way to consolidate related functions into a single JS file. For example, let's say I wanted to create 5 different template functions that each produce a different text string to inject. Normally in JS, I would create a simple object with each key being a function and finally export the object.

```javascript
const SomeLib = {
  functionOne() {
    // does something
  },
  functionTwo() { },
  functionThree() { },
  functionFour() { },
  functionFive() { }
}

module.exports = SomeLib;
```
Then I would be able to reference the desired function with `SomeLib.functionOne()`, etc.

Is there any way to do something similar in Templater where I can use `<% tp.user.SomeLib.functionOne() %>`?

**Comment by welpdx** [comment, marked as answer]
What I did is create a class and export the class in the .js file. Then, in templater, I create a new class, and refer to functions from that class. 

<details ><summary>filename.js</summary>
<p>

```js
// filename.js
class SomeLib {
  functionOne() {new Notice("functionOne", 5000)}
  functionTwo() {new Notice("functionTwo", 5000)}
  functionThree() {new Notice("functionThree", 5000)}
  functionFour() {new Notice("functionFour", 5000)}
  functionFive() {new Notice("functionFive", 5000)}
}
module.exports = SomeLib
```

</p>
</details>




<details><summary>Template</summary>
<p>

```js
// Templater Template
<%* 
const f = new tp.user.filename;
f.functionTwo();
%>
```

</p>
</details>


<details>
           <summary style="font-size:20">Demo</summary>
           <img src="https://i.imgur.com/jaRqwq6.gif" alt="demo" />
</details>


Hope that helps.

~Welp

**Comment by dainbrump** [comment]
@welpdx Interesting solution. Of course, it's the one approach I had not tried. 😏 Thanks for the suggestion.

**Comment by welpdx** [comment]
@dainbrump, no problem! 😃 
That was a really good question too! I haven't even touched the Templater User Script settings because I, too, thought you could only export one js function at a time. You question really helped me explore the possibility of exporting multiple functions, and I think I might use this down the line lol

**Comment by dainbrump** [comment]
@SilentVoid13 Unless there is an "official" answer or solution to this, I'll gladly accept the answer @welpdx gave and we can close this. Hopefully this might help others. Now to setup a typescript project to hold all my fancy new user scripts. :-D

**Comment by AB1908** [comment]
You may also find the CustomJS plugin useful. You can use that library of functions elsewhere like in DataviewJS.

**Comment by butleger** [comment]
I think it would be really convenient to export multiple functions from file in object style, may be reopen issue as enhancment?

**Reply by McGlear** [reply]
Still merely a workaround, but you can of course also export a function that `return`s an object that you define within the user function script, then from your templates call
`let obj = tp.user.getObject()`

With the user function something like
 ```
function getObj() {
  return {name: "Simple object"}
}
module.exports = getObj
```
